packages/vite/src/node/build.ts

```diff
@@ -542,55 +544,65 @@ async function doBuild(
     parallelBuilds.push(bundle)
 
     const generate = (output: OutputOptions = {}) => {
-      return bundle[options.write ? 'write' : 'generate'](
-        buildOutputOptions(output)
-      )
+      return bundle[options.write ? 'write' : 'generate'](output)
     }
 
     if (options.write) {
-      prepareOutDir(outDir, options.emptyOutDir, config)
+      prepareOutDir(outDirs, options.emptyOutDir, config)
     }
 
-    if (Array.isArray(outputs)) {
-      const res = []
-      for (const output of outputs) {
-        res.push(await generate(output))
-      }
-      return res
-    } else {
-      return await generate(outputs)
+    const res = []
+    for (const output of normalizedOutputs) {
+      res.push(await generate(output))
     }
+    return Array.isArray(outputs) ? res : res[0]
   } catch (e) {
     outputBuildError(e)
     throw e
   }
 }
 
 function prepareOutDir(
-  outDir: string,
+  outDirs: string[],
   emptyOutDir: boolean | null,
   config: ResolvedConfig
 ) {
-  if (fs.existsSync(outDir)) {
-    if (
-      emptyOutDir == null &&
-      !normalizePath(outDir).startsWith(config.root + '/')
-    ) {
-      // warn if outDir is outside of root
-      config.logger.warn(
-        colors.yellow(
-          `\n${colors.bold(`(!)`)} outDir ${colors.white(
-            colors.dim(outDir)
-          )} is not inside project root and will not be emptied.\n` +
-            `Use --emptyOutDir to override.\n`
+  for (const outDir of outDirs) {
+    if (fs.existsSync(outDir)) {
+      if (
+        emptyOutDir == null &&
+        !normalizePath(outDir).startsWith(config.root + '/')
+      ) {
+        // warn if outDir is outside of root
+        config.logger.warn(
+          colors.yellow(
+            `\n${colors.bold(`(!)`)} outDir ${colors.white(
+              colors.dim(outDir)
+            )} is not inside project root and will not be emptied.\n` +
+              `Use --emptyOutDir to override.\n`
+          )
         )
-      )
-    } else if (emptyOutDir !== false) {
-      emptyDir(outDir, ['.git'])
+      } else if (emptyOutDir !== false) {
+        // skip those other outDirs which are nested in current outDir
+        const skipDirs = outDirs
+          .map((dir) => {
+            const relative = path.relative(outDir, dir)
+            if (
+              relative &&
+              !relative.startsWith('..') &&
+              !path.isAbsolute(relative)
+            ) {
+              return relative
+            }
+            return ''
+          })
+          .filter(Boolean)
+        emptyDir(outDir, [...skipDirs, '.git'])
```

Comment:
I don't quite understand this part. If we're already emptying all `outDirs` do we need to skip those that are nested? What about if we filter out the dirs that are nested since we know a higher-up directory would also remove it too. Something like:

```js
outDirs = outDirs.filter((dir) => {
  return outDirs.every((outDir) => !dir.startsWith(outDir))
})
```

at the start of this function.

Reply:
It is to ensure the `.git` dir not be removed in each outDir.

Currently we improved the `emptyDir` to support skip arbitrary level of nesting paths, I think this part can be improved first collect all of `.git` dirs under a higher outputDir to decrease loop count.

But the count of outDirs may 3 or 4 at most (in normal case), maybe the improvement is not necessary.

Reply:
I think it isn't only about the `.git` file, but it is also good that these nested `outDirs` aren't deleted. Emptying should remove all generated dist files, but not the dir structure that was created to hold them.

Reply:
Ah alright I see now. I guess my concern is that `emptyDir` should be a simple concept, but we're adding a lot logic to reach that for an edgecase. Do we also support nesting out dirs? That sounds like it would cause other side effects too.

I also realized that this decision was made at [this comment](https://github.com/vitejs/vite/pull/9748#discussion_r950061172), I feel like it's not something worth supporting now, but the work has been done 😬 

Anyways I don't mind if we do go ahead with what we have now, just that I feel like we're shipping more code that most won't be using.

Reply:
I think that there is no harm in having emptyDir support this (the added code isn't that much), and as you say the work has already been done. But let's bring @sapphi-red to get another vote. I also don't mind deciding that nested out dirs will be wiped out for now, and we can add back this logic if someone presents the use case later.

Reply:
I think it's ok to add this as the complexity is not too much.